extensions/amp-viewer-integration/0.1/amp-viewer-integration.js

```diff
@@ -14,5 +14,85 @@
  * limitations under the License.
  */
 
+import {Messaging} from './messaging.js';
+import {listen} from '../../../src/event-helper';
+import {viewerForDoc} from '../../../src/viewer';
+import {dev,user} from '../../../src/log';
+
+const TAG = 'amp-viewer-integration';
+
+/**
+ * @fileoverview This is the communication protocol between AMP and the viewer.
+ * This should be included in an AMP html file to communicate with the viewer.
+ */
 export class AmpViewerIntegration {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @const {!Window} win */
+    this.win = win;
+  }
+
+  /**
+   * Initiate the handshake. If handshake confirmed, start listening for
+   * messages.
+   * @return {!Promise|undefined}
+   */
+  init() {
+    dev().info('AVI', TAG + ' => handshake init()');
+    const viewer = viewerForDoc(this.win.document);
+    return this.getHandshakePromise_(viewer)
+      .then(viewerOrigin => {
+        dev().info('AVI', TAG + ' => listening for messages');
+        const messaging =
+          new Messaging(this.win, this.win.parent, viewerOrigin,
+            (type, payload, awaitResponse) => {
+              return viewer.receiveMessage(
+                type, /** @type {!JSONType} */ (payload), awaitResponse);
+            });
+        viewer.setMessageDeliverer((type, payload, awaitResponse) => {
+          return messaging.sendRequest(type, payload, awaitResponse);
+        }, viewerOrigin);
+      });
+  }
+
+  /**
+   * Send a handshake request, and listen for a handshake response to
+   * confirm the handshake.
+   * @param {!../../../src/service/viewer-impl.Viewer} viewer
+   * @return {!Promise}
+   * @private
+   */
+  getHandshakePromise_(viewer) {
+    const win = this.win;
+    return new Promise(resolve => {
+      const unconfirmedViewerOrigin =
+        viewer.getParam('viewerorigin') || viewer.getParam('origin');
```

Comment:
Sorry, I think my request was unclear. 

> Please add another || with this param here: https://github.com/ampproject/amphtml/blob/master/src/service/viewer-impl.js#L233

I meant that we need to test against `viewerorigin` inside `viewer-impl.js` (see the link), not against `origin` here.